In [102]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [103]:
PathTrain = 'E-Commerce_train.xlsx'
PathTest = 'E-Commerce_test.xlsx'

VALORES DE CADA CAMPO 

ID: identificador del registro de orden (valor entero).
Warehouse_block: Almacén de distribución de donde salió la orden (A a F).
Mode_of_Shipment: Medio de transporte (Flight, Road, Ship).
Customer_care_calls: Número de llamadas a atención al cliente que hubo por esa orden. (valores enteros del 2 al 7)
Customer_rating: Puntaje del cliente (valores enteros 1 al 5).
Cost_of_the_Product: Costo del producto (valor numérico entero de 96 a 310).
Prior_purchases: Número de compras previas realizadas por el cliente (valor numérico entero de 2 a 10).
Product_importance: Nivel de importancia del producto (low, medium, high).
Gender: Género del comprador (F, M).
Discount_offered: Porcentaje de descuento ofrecido por esa compra (valor numérico entero de 1 a 65):
Weight_in_gms: Peso del paquete de la orden, en gramos (valor numérico entero de 1001 a 7846).
Reached.on.Time_Y.N: Información sobre la llegada del paquete a destino (1 si llegó a tiempo, 0 si no llegó a tiempo).

In [104]:
#Cargamos el archivo de entrenamiento formato excel a Dataframe
DataTrain = pd.read_excel(PathTrain, sheet_name = 'Sheet1')
DataTrain

ID Warehouse_block Mode_of_Shipment  Customer_care_calls  \
0        1               D           Flight                    4   
1        2               F           Flight                    4   
2        3               A           Flight                    2   
3        4               B           Flight                    3   
4        5               C           Flight                    2   
...    ...             ...              ...                  ...   
8994  8995               D             Ship                    3   
8995  8996               F             Ship                    5   
8996  8997               A             Ship                    4   
8997  8998               B             Ship                    4   
8998  8999               C             Ship                    5   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
0                   2                  177                3   
1                   5                  216                2   
2                   2                  183                4   
3                   3                  176                4   
4                   2                  184                3   
...               ...                  ...              ...   
8994                1                  217                3   
8995                3                  232                3   
8996                5                  260                3   
8997                2                  184                3   
8998                4                  145                4   

     Product_importance Gender  Discount_offered  Weight_in_gms  \
0                   low      F                44           1233   
1                   low      M                59           3088   
2                   low      M                48           3374   
3                medium      M                10           1177   
4                medium      F                46           2484   
...                 ...    ...               ...            ...   
8994                low      F                 1           4177   
8995             medium      M                 3           4526   
8996             medium      F                 6           4221   
8997             medium      F                 5           5931   
8998                low      M                 4           4790   

      Reached.on.Time_Y.N  
0                       1  
1                       1  
2                       1  
3                       1  
4                       1  
...                   ...  
8994                    1  
8995                    0  
8996                    0  
8997                    1  
8998                    0  

[8999 rows x 12 columns]

In [105]:
#Cambiamos los tipos de variables de las columnas con campos que contienen string a numericos.
DataTrain.Mode_of_Shipment = DataTrain.Mode_of_Shipment.replace({'Road': 1 , 'Flight': 2 , 'Ship' : 3})
DataTrain.Product_importance = DataTrain.Product_importance.replace({'low': 1 , 'medium': 2 , 'high' : 3})
DataTrain.Warehouse_block = DataTrain.Warehouse_block.replace({'A': 1 , 'B': 2 , 'C' : 3 , 'D': 4 , 'E': 5 , 'F' : 6})
DataTrain

ID  Warehouse_block  Mode_of_Shipment  Customer_care_calls  \
0        1                4                 2                    4   
1        2                6                 2                    4   
2        3                1                 2                    2   
3        4                2                 2                    3   
4        5                3                 2                    2   
...    ...              ...               ...                  ...   
8994  8995                4                 3                    3   
8995  8996                6                 3                    5   
8996  8997                1                 3                    4   
8997  8998                2                 3                    4   
8998  8999                3                 3                    5   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
0                   2                  177                3   
1                   5                  216                2   
2                   2                  183                4   
3                   3                  176                4   
4                   2                  184                3   
...               ...                  ...              ...   
8994                1                  217                3   
8995                3                  232                3   
8996                5                  260                3   
8997                2                  184                3   
8998                4                  145                4   

      Product_importance Gender  Discount_offered  Weight_in_gms  \
0                      1      F                44           1233   
1                      1      M                59           3088   
2                      1      M                48           3374   
3                      2      M                10           1177   
4                      2      F                46           2484   
...                  ...    ...               ...            ...   
8994                   1      F                 1           4177   
8995                   2      M                 3           4526   
8996                   2      F                 6           4221   
8997                   2      F                 5           5931   
8998                   1      M                 4           4790   

      Reached.on.Time_Y.N  
0                       1  
1                       1  
2                       1  
3                       1  
4                       1  
...                   ...  
8994                    1  
8995                    0  
8996                    0  
8997                    1  
8998                    0  

[8999 rows x 12 columns]

In [106]:
#No tenemos en cuenta las columnas 'ID' y 'Genger' de los atributos ya que no aportan influencia en el resultado de manera directa. 
X = DataTrain[['Warehouse_block','Mode_of_Shipment', 'Customer_care_calls', 'Customer_rating','Cost_of_the_Product','Prior_purchases','Product_importance','Discount_offered' , 'Weight_in_gms']]
y = DataTrain[['Reached.on.Time_Y.N']]

In [107]:
#Importamos el separador
from sklearn.model_selection import train_test_split

#Separamos el %30 de los datos para el test y lo restante lo entrenamos
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=0)

In [108]:
from sklearn.tree import DecisionTreeClassifier

# Instanciamos un objeto de la clase DecisionTreeClassifier

Arbol = DecisionTreeClassifier(max_depth = 4, random_state = 0) 

In [109]:
# Entrenamos el modelo

Arbol.fit(X_train.values,y_train.values)

DecisionTreeClassifier(max_depth=4, random_state=0)

In [110]:
# Predecimos los atributos apartados para el testeo

y_pred = Arbol.predict(X_test.values)

In [111]:
from sklearn.metrics import accuracy_score , recall_score

Accuracy = accuracy_score(y_pred,y_test)
Recall = recall_score(y_pred,y_test)

print(Accuracy , Recall)


0.702962962962963 0.9607250755287009


Según pruebas cambiando la profundidad: 
3 - (0.6981481481481482 0.8350668647845468)
4 - (0.702962962962963 0.9607250755287009) 
5 - (0.7048148148148148 0.9581673306772909)
6 - (0.7025925925925925 0.8733660130718954)
10 - (0.6885185185185185 0.7971506105834464)

Agarramos como profundidad 4 ya que es el que mayor Recall nos dió en el Subset de entrenamiento.

In [112]:
#Cargamos el archivo de Test formato excel a Dataframe
DataTest = pd.read_excel(PathTest, sheet_name = 'Sheet1')
DataTest 

ID Warehouse_block Mode_of_Shipment  Customer_care_calls  \
0      9000               F             Ship                    5   
1      9001               D             Ship                    5   
2      9002               F             Ship                    3   
3      9003               A             Ship                    5   
4      9004               B             Ship                    4   
...     ...             ...              ...                  ...   
1995  10995               A             Ship                    4   
1996  10996               B             Ship                    4   
1997  10997               C             Ship                    5   
1998  10998               F             Ship                    5   
1999  10999               D             Ship                    2   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
0                   1                  189                2   
1                   3                  150                2   
2                   3                  200                3   
3                   1                  133                2   
4                   2                  260                3   
...               ...                  ...              ...   
1995                1                  252                5   
1996                1                  232                5   
1997                4                  242                5   
1998                2                  223                6   
1999                5                  155                5   

     Product_importance Gender  Discount_offered  Weight_in_gms  
0                medium      M                10           5248  
1                  high      F                 9           4446  
2                   low      M                 3           5739  
3                medium      M                 1           4643  
4                   low      M                 4           5504  
...                 ...    ...               ...            ...  
1995             medium      F                 1           1538  
1996             medium      F                 6           1247  
1997                low      F                 4           1155  
1998             medium      M                 2           1210  
1999                low      F                 6           1639  

[2000 rows x 11 columns]

In [113]:
#Aplicamos las mismas transformaciones al Dataframe de Test 

DataTest.Mode_of_Shipment = DataTest.Mode_of_Shipment.replace({'Road': 1 , 'Flight': 2 , 'Ship' : 3})
DataTest.Product_importance = DataTest.Product_importance.replace({'low': 1 , 'medium': 2 , 'high' : 3})
DataTest.Warehouse_block = DataTest.Warehouse_block.replace({'A': 1 , 'B': 2 , 'C' : 3 , 'D': 4 , 'E': 5 , 'F' : 6})
DataTest

ID  Warehouse_block  Mode_of_Shipment  Customer_care_calls  \
0      9000                6                 3                    5   
1      9001                4                 3                    5   
2      9002                6                 3                    3   
3      9003                1                 3                    5   
4      9004                2                 3                    4   
...     ...              ...               ...                  ...   
1995  10995                1                 3                    4   
1996  10996                2                 3                    4   
1997  10997                3                 3                    5   
1998  10998                6                 3                    5   
1999  10999                4                 3                    2   

      Customer_rating  Cost_of_the_Product  Prior_purchases  \
0                   1                  189                2   
1                   3                  150                2   
2                   3                  200                3   
3                   1                  133                2   
4                   2                  260                3   
...               ...                  ...              ...   
1995                1                  252                5   
1996                1                  232                5   
1997                4                  242                5   
1998                2                  223                6   
1999                5                  155                5   

      Product_importance Gender  Discount_offered  Weight_in_gms  
0                      2      M                10           5248  
1                      3      F                 9           4446  
2                      1      M                 3           5739  
3                      2      M                 1           4643  
4                      1      M                 4           5504  
...                  ...    ...               ...            ...  
1995                   2      F                 1           1538  
1996                   2      F                 6           1247  
1997                   1      F                 4           1155  
1998                   2      M                 2           1210  
1999                   1      F                 6           1639  

[2000 rows x 11 columns]

In [114]:
#No tenemos en cuenta las columnas 'ID' y 'Genger' de los atributos ya que no aportan influencia en el resultado de manera directa. 
XTest = DataTest[['Warehouse_block','Mode_of_Shipment', 'Customer_care_calls', 'Customer_rating','Cost_of_the_Product','Prior_purchases','Product_importance','Discount_offered' , 'Weight_in_gms']]

In [121]:
#Predecimos usando los valores nuevos de los atributos y el objeto anteriormente creado 'Arbol'
Pred = Arbol.predict(XTest) 

/home/justo99/.local/lib/python3.8/site-packages/sklearn/base.py:443: UserWarning: X has feature names, but DecisionTreeClassifier was fitted without feature names
  warnings.warn(


In [122]:
pred = pd.DataFrame(Pred)

In [123]:
#Creamos el archivo csv, utilizando nombre de github como raíz
pred.to_csv('JustoCosta.csv' , sep = ',', index = False)